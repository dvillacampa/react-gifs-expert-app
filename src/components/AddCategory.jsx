import { useState } from 'react';
import { PropTypes} from 'prop-types';


export const AddCategory = ({ onNewCategory }) => {

    // El valor inicial del Hook es vacío. 
    // La función setInputValue es la responsable de establecer el valor del <input> en inputValue
    // inputValue guarda el estado del <input> es decir, el contenido visible del <input>
    const [inputValue, setInputValue] = useState(''); 

    // La función onChange recibe como argumento event
    // Se usa este argumento para llamar a la función del Hook onNewCategory y actualizar el estado del componente 'useState' desde el componente padre.

    // Se podría haber hecho así:
    // onChange = { (event) => onInputChange( event ) }

    // Desestructuramos 'event' en 'target' para reducir el código.
    // El elemento de la jerarquía 'target' contine el valor del <input> 'event.target.value'
    const onInputChange = ( { target } ) => {
        setInputValue( target.value ); // Llama a la función callback setInputValue con el nuevo valor del <input>
    };

    const onFormSubmit = ( event ) => {
        event.preventDefault();
        const newCategory = inputValue.trim();
        // Si ha escrito menos de 2 caracteres no hace nada aunque se haya enviado el form.
        if( newCategory.length <= 1) return;

        // setCategories( categories => [ ...categories, inputValue ] ); Esto lo usamos cuando recibíamos el Array de Categorías del padre

        // Vamos a llamar al evento que hemos recibido como props desde el componente padre <GifExpertApp> al añadir este componente
        onNewCategory( newCategory );
        // Elimina el texto del Input
        setInputValue('');
    }

    // Se define un form para responder el evento onSubmit con la tecla "Enter"
    // El <form> hace de contenedor del elemento <input>.
    // Properties:
    //      paceholder  = texto que se muestra cuando el <input> aparece vacío, value = vacío
    //      value       = valor que se muestra en el <input>
    //      type        = comportamiento del <input>
    //      onChange    = un evento que necesitamos manejar para reflejar los cambios que hace el usuario en el <input>
    return (
            <form 
                onSubmit={ onFormSubmit }
                aria-label="form"> 
                <input 
                    type="text"
                    placeholder="Buscar gifs"
                    value={ inputValue }
                    onChange={ onInputChange }
                />
            </form>
    );
}

AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired,
}