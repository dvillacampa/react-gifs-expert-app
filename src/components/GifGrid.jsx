import { useFetchGifs } from '../hooks';
import { GifItem } from '../components';
import { PropTypes} from 'prop-types';

export const GifGrid = ( {category} ) => {

    const { images, isLoading } = useFetchGifs( category);

    return (
        <>
            <h3>{category}</h3>
            {   // So isLoading es false no renderiza el elemento h2
            }
            { isLoading && <h2>Cargando...</h2> } 
            <div className='card-grid'>
            {
                images.map ( ( imagen )  => 
                    <GifItem key={ imagen.id } 
                        { ...imagen } /> // Expande todas las propiedas/atributos del objeto imagen. Forma f√°cil de pasar muchos argumentos
                    
                )
            }
            </div>
        </>
    )

};

GifGrid.propTypes =  {
    category: PropTypes.string.isRequired
}