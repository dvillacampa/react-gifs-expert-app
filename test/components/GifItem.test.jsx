import { render, screen } from '@testing-library/react';
import { GifItem } from '../../src/components/GifItem';

describe('Prueba del componente <GifItem>', () => {

    const title = 'Zelda';
    const url = 'http://url.de.prueba.com/';

    test('Debería hacer match con el snapshot', () => {

        const { container } = render(<GifItem title={ title } url={ url }/>);

        expect( container ).toMatchSnapshot();

    });

    test('La propiedad src del elemento <img> debería coincidir con la URL de la imagen', () => {

        render(<GifItem title={ title } url={ url }/>);
        expect( screen.getByRole( 'img' ).src).toBe( url );
    });

    test('La propiedad alt del elemento <img> debería coincidir con el título de la imagen', () => {

        render(<GifItem title={ title } url={ url }/>);
        const { alt } = screen.getByRole( 'img' );
        expect( alt ).toBe( title );

    });

    test('Debe renderizar el texto title', () => {

        render(<GifItem title={ title } url={ url }/>);
        expect( screen.getByText( title )).toBeTruthy();

    });


});