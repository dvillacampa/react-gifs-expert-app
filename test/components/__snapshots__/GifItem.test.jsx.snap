// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del componente <GifItem> Deber√≠a hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Zelda"
      src="http://url.de.prueba.com/"
    />
    <p>
      Zelda
    </p>
  </div>
</div>
`;
